src/shm/pool/multi.rs

```diff
@@ -0,0 +1,397 @@
+//!             self.surface.1 = i;
+//!             if let Ok((offset, buffer, slice)) = self.pool.create_buffer(
+//!                 100,
+//!                 100 * 4,
+//!                 100,
+//!                 &self.surface,
+//!                 Format::Argb8888,
+//!                 conn,
+//!             ) {
+//!                 /*
+//!                     insert drawing code here
+//!                 */
+//!                 surface.attach(conn, Some(buffer), 0, 0);
+//!                 surface.commit(conn);
+//!                 // We exit the function after the draw.
+//!                 return;
+//!             }
+//!         }
+//!         /*
+//!             If there's no buffer available we can for example request a frame callback
+//!             and trigger a redraw when it fires.
+//!             (not shown in this example)
+//!         */
+//!     }
+//! }
+//!
+//! fn draw(slice: &mut [u8]) {
+//!     todo!()
+//! }
+//!
+//! ```
+//!
+
+use std::io;
+
+use std::sync::{
+    atomic::{AtomicBool, Ordering},
+    Arc,
+};
+use wayland_client::{
+    protocol::{wl_buffer, wl_shm, wl_shm_pool},
+    ConnectionHandle, Proxy, WEnum,
+};
+
+use super::raw::RawPool;
+
+#[derive(Debug, thiserror::Error)]
+pub enum PoolError {
+    #[error("buffer is currently used")]
+    InUse,
+    #[error("buffer is overlapping another")]
+    Overlap,
+    #[error("buffer could not be found")]
+    NotFound,
+}
+
+/// This pool manages buffers associated with keys.
+/// Only one buffer can be attributed to a given key.
+#[derive(Debug)]
+pub struct MultiPool<K: Clone + PartialEq> {
+    buffer_list: Vec<BufferSlot<K>>,
+    pub(crate) inner: RawPool,
+}
+
+#[derive(Debug, thiserror::Error)]
+struct BufferSlot<K: Clone + PartialEq> {
+    free: Arc<AtomicBool>,
+    size: usize,
+    used: usize,
+    offset: usize,
+    buffer: Option<wl_buffer::WlBuffer>,
+    key: K,
+}
+
+impl<K: Clone + PartialEq> MultiPool<K> {
+    pub(crate) fn new(inner: RawPool) -> Self {
+        Self { inner, buffer_list: Vec::new() }
+    }
+
+    /// Resizes the memory pool, notifying the server the pool has changed in size.
+    ///
+    /// The wl_shm protocol only allows the pool to be made bigger. If the new size is smaller than the
+    /// current size of the pool, this function will do nothing.
+    pub fn resize(&mut self, size: usize, conn: &mut ConnectionHandle) -> io::Result<()> {
+        self.inner.resize(size, conn)
+    }
+
+    /// Removes the buffer with the given key from the pool and rearranges the others.
+    pub fn remove<Q>(&mut self, key: &Q, conn: &mut ConnectionHandle)
+    where
+        Q: Eq,
+        K: std::borrow::Borrow<Q>,
+    {
+        if let Some((i, buf)) =
+            self.buffer_list.iter().enumerate().find(|(_, slot)| slot.key.borrow().eq(key))
+        {
+            let mut offset = buf.offset;
+            if let Some(buf) = self.buffer_list.remove(i).buffer {
+                buf.destroy(conn);
```

Comment:
This may destroy a buffer in use

Reply:
I think this if let block can be replaced by
```rust
self.buffer_list.remove(i);
```
The buffer will not be destroyed so it might not be ideal.

I think you would only remove a buffer if you were done with it ie the surface is destroyed. In that case, I don't think it's a big deal to potentially destroy the buffer since the compositor should not display anything on it in the following frame.